File: python_mods/Algorithms/RANSAC.cpp
std::tuple<Eigen::Vector4d, std::vector<size_t>> PointCloud::SegmentPlane(
        const double distance_threshold /* = 0.01 */,
        const int ransac_n /* = 3 */,
        const int num_iterations /* = 100 */,
        const double probability /* = 0.99999999 */,
        utility::optional<int> seed /* = utility::nullopt */) const {
    if (probability <= 0 || probability > 1) {
        utility::LogError("Probability must be > 0 or <= 1.0");
    }

    RANSACResult result;

    // Initialize the best plane model.
    Eigen::Vector4d best_plane_model = Eigen::Vector4d(0, 0, 0, 0);

    size_t num_points = points_.size();
    RandomSampler<size_t> sampler(num_points, seed);

    // Return if ransac_n is less than the required plane model parameters.
    if (ransac_n < 3) {
        utility::LogError(
                "ransac_n should be set to higher than or equal to 3.");
        return std::make_tuple(Eigen::Vector4d(0, 0, 0, 0),
                               std::vector<size_t>{});
    }
    if (num_points < size_t(ransac_n)) {
        utility::LogError("There must be at least 'ransac_n' points.");
        return std::make_tuple(Eigen::Vector4d(0, 0, 0, 0),
                               std::vector<size_t>{});
    }

    // Use size_t here to avoid large integer which acceed max of int.
    size_t break_iteration = std::numeric_limits<size_t>::max();
    int iteration_count = 0;

#pragma omp parallel for schedule(static)
    for (int itr = 0; itr < num_iterations; itr++) {
        if ((size_t)iteration_count > break_iteration) {
            continue;
        }

        const std::vector<size_t> sampled_indices = sampler(ransac_n);
        std::vector<size_t> inliers = sampled_indices;

        // Fit model to num_model_parameters randomly selected points among the
        // inliers.
        Eigen::Vector4d plane_model;
        if (ransac_n == 3) {
            plane_model = TriangleMesh::ComputeTrianglePlane(
                    points_[inliers[0]], points_[inliers[1]],
                    points_[inliers[2]]);
        } else {
            plane_model = GetPlaneFromPoints(points_, inliers);
        }

        if (plane_model.isZero(0)) {
            continue;
        }

        double error = 0;
        inliers.clear();
        auto this_result = EvaluateRANSACBasedOnDistance(
                points_, plane_model, inliers, distance_threshold, error);
#pragma omp critical
        {
            if (this_result.fitness_ > result.fitness_ ||
                (this_result.fitness_ == result.fitness_ &&
                 this_result.inlier_rmse_ < result.inlier_rmse_)) {
                result = this_result;
                best_plane_model = plane_model;
                if (result.fitness_ < 1.0) {
                    break_iteration = std::min(
                            log(1 - probability) /
                                    log(1 - pow(result.fitness_, ransac_n)),
                            (double)num_iterations);
                } else {
                    // Set break_iteration to 0 to force to break the loop.
                    break_iteration = 0;
                }
            }
            iteration_count++;
        }
    }

    // Find the final inliers using best_plane_model.
    std::vector<size_t> final_inliers;
    if (!best_plane_model.isZero(0)) {
        for (size_t idx = 0; idx < points_.size(); ++idx) {
            Eigen::Vector4d point(points_[idx](0), points_[idx](1),
                                  points_[idx](2), 1);
            double distance = std::abs(best_plane_model.dot(point));

            if (distance < distance_threshold) {
                final_inliers.emplace_back(idx);
            }
        }
    }

    // Improve best_plane_model using the final inliers.
    best_plane_model = GetPlaneFromPoints(points_, final_inliers);

    utility::LogDebug(
            "RANSAC | Inliers: {:d}, Fitness: {:e}, RMSE: {:e}, Iteration: "
            "{:d}",
            final_inliers.size(), result.fitness_, result.inlier_rmse_,
            iteration_count);
    return std::make_tuple(best_plane_model, final_inliers);
}

File: 3D_Object_Recognition/Algorithms_From_Scratch/RANSAC.py
import numpy as np
from typing import Optional, Tuple


def SegmentPlane(distance_threshold: float = 0.01,
        ransac_n: int = 3,
        num_iterations: int = 100,
        probability: float = 0.99999999,
        seed: Optional[int] = None) -> Tuple[np.ndarray, np.ndarray]:
    if probability <= 0 or probability > 1:
        print("Probability must be > 0 or <= 1.0")
        raise ValueError

    # RANSACResult result;

    # Initialize the best plane model.
    best_plane_model = np.array([0, 0, 0, 0])

    num_points = self.points_.size
    sampler = RandomSampler(num_points, seed)

    # Return if ransac_n is less than the required plane model parameters.
    if ransac_n < 3:
        print("ransac_n should be set to higher than or equal to 3.")
        return np.array([0, 0, 0, 0]), np.ndarray([])

    if num_points < ransac_n:
        print("There must be at least 'ransac_n' points.")
        return np.array([0, 0, 0, 0]), np.ndarray([])

    # Use size_t here to avoid large integer which acceed max of int.
    # size_t break_iteration = std::numeric_limits<size_t>::max();
    break_iteration = 1e32
    iteration_count = 0

    for itr in range(num_iterations):
        if iteration_count > break_iteration:
            continue

        sampled_indices = sampler(ransac_n)  ## O(n)
        inliers = sampled_indices

        # Fit model to num_model_parameters randomly selected points among the inliers.
        # Eigen::Vector4d plane_model;
        if ransac_n == 3:
            plane_model = TriangleMesh.ComputeTrianglePlane(self.points_[inliers[0]], self.points_[inliers[1]], self.points_[inliers[2]]);
        else:
            plane_model = GetPlaneFromPoints(self.points_, inliers)  # O(n)

        if plane_model.isZero(0):
            continue

        error = 0
        inliers.clear()
        this_result = EvaluateRANSACBasedOnDistance(points_, plane_model, inliers, distance_threshold, error)
        if this_result.fitness_ > result.fitness_ or (this_result.fitness_ == result.fitness_ and this_result.inlier_rmse_ < result.inlier_rmse_):
            result = this_result
            best_plane_model = plane_model
            if result.fitness_ < 1.0:
                break_iteration = min(np.log(1 - probability) / np.log(1 - pow(result.fitness_, ransac_n)), num_iterations)
            else:
                # Set break_iteration to 0 to force to break the loop.
                break_iteration = 0

        iteration_count += 1


    # Find the final inliers using best_plane_model.
    # std::vector<size_t> final_inliers;
    if not best_plane_model.isZero(0):
        for idx in range(self.points_.size()):
            point = np.array([self.points_[idx](0), self.points_[idx](1), self.points_[idx](2), 1])
            distance = abs(best_plane_model.dot(point))

            if distance < distance_threshold:
                final_inliers.emplace_back(idx)

    # Improve best_plane_model using the final inliers.
    best_plane_model = GetPlaneFromPoints(points_, final_inliers)

    print("RANSAC | Inliers: {:d}, Fitness: {:e}, RMSE: {:e}, Iteration: {:d}", final_inliers.size(), result.fitness_, result.inlier_rmse_, iteration_count)
    return best_plane_model, final_inliers